core/src/main/ruby/jruby/kernel/enumerator.rb

```diff
@@ -7,6 +7,100 @@ class Yielder
     # or http://bugs.ruby-lang.org/issues/7696
     attr_accessor :backports_memo
   end
+  def next
+    return @__lookahead__.shift unless @__lookahead__.empty?
+
+    unless @__generator__
+      # Allow #to_generator to return nil, indicating it has none for
+      # this method.
+      if @object.respond_to? :to_generator
+        @__generator__ = @object.to_generator(@iter)
+      end
+
+      if !@__generator__
+        @__generator__ = FiberGenerator.new(self)
+      end
+    end
+
+    begin
+      return @__generator__.next if @__generator__.next?
+    rescue StopIteration
+      nil # the enumerator could change between next? and next leading to StopIteration
+    end
+
+    exception = StopIteration.new 'iteration reached an end'
+    JRuby.ref(exception).result = @__generator__.result
```

Comment:
a bit ugly maybe worth extending `StopIteration.new` 
or having a `StopIteration.__new__('msg', result)`?
... you wouldn't need to have `RubyStopIteration#setResult` around

Reply:
Yeah this is troublesome. The result is not supposed to be mutable through normal Ruby APIs, if we want to be strictly compatible, but there's also no way to construct a StopIteration with a specific result from Ruby. Both the Rubinius and TruffleRuby versions of this code do something basically the same. I went with what was simplest for now.